### ISSUE TITLE
Loss of messages

### ISSUE THREAD

**Issue by frankboehmer** [RIAEvangelist/node-ipc#145]
node-ipc will lose/mix up messages when dealing with long messages and heavy concurrency:
socketServer.js / gotData holds a buffer this.ipcBuffer. This variable is shared with all client connections. Thus, concurrent clients will fill up the same buffer, messing up things.
A quick but dirty fix might be to change this to socket.ipcBuffer.
A better option might be to have it as a closure in serverCreated() {...}

**Comment by frankboehmer** [comment]
Would be great if someone could fix this or merge my pull request. This is a serious flaw and may affect unnoticed a lot of production systems.

**Comment by frank-dspeed** [comment]
@frankboehmer when some one uses this in production he is doomed anyway :dagger:

**Comment by TanninOne** [comment]
@frank-dspeed Could you elaborate a bit? I'm also noticing messages going missing even with just a single client but is this module known to be too buggy/unsafe?
Is there a more robust alternative?

**Comment by frankboehmer** [comment]
@TanninOne the package is not maintained and has critical issues. 
(Latest commit 70e03c1  on 26 Oct 2017)

**Comment by TanninOne** [comment]
Well, an ipc library is a narrow scope, I'm actually fine with a library like that being done and not needing constant updates.
The critical issues that apparently don't get fixed are a bigger problem though.
Soo, any good alternatives?

**Comment by windware-ono** [comment]
Surprised I couldn't find any viable alternatives to do simple IPC.

I was actually using [axon](https://www.npmjs.com/package/axon) before but I didn't feel the code was too complete (client connections somehow didn't close and callback not getting called.) and then found node-ipc has a serious flaw...

But figured a simple 'net' would do it.

server.js
``` js
const server = require('net').createServer((stream) =>
{
	let body = ''

	stream.on('data', (chunk) =>
	{
		body += chunk
	})

	stream.on('end', () =>
	{
		const message = JSON.parse(body.toString())
		console.log(`Received : ${JSON.stringify(message)}`)
	})

	stream.on('error', (error) =>
	{
		throw error
	})
})

server.on('listening', () =>
{
	console.log('Server listening')
})

server.on('error', (error) =>
{
	throw error
})

server.listen('/tmp/nodejs-ipc.sock')

//Should also listen on 'uncaughtException' and others to clean up the socket file
process.on('SIGINT', () =>
{
	server.close()
})
```

client.js
``` js
const client = require('net').createConnection('/tmp/nodejs-ipc.sock')

client.on('ready', () =>
{
	client.write(JSON.stringify({some: 'abc', vars: 123}))
	client.end()
})

client.on('error', (error) =>
{
	throw error
})
```

**Comment by RIAEvangelist** [comment]
This will be fixed in v10.0.0 it was a collision issue.

**Comment by joshuaalayon** [comment]
Be aware - this package was found to contain malware that detects its users machines purported geographical location and replaces the data from the machine if it does not like the location it detects

If you are still using this package - you may do better by locking it to an older version or, preferably, replacing it altogether along with all software maintained by the repos maintainer(s)

https://www.bleepingcomputer.com/news/security/big-sabotage-famous-npm-package-deletes-files-to-protest-ukraine-war/amp/
